### PR TITLE
[optimizer] refactor Adam to use functional API

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45221 [dist_optim] introduce distributed functional optimizer
* **#44791 [optimizer] refactor Adam to use functional API**
* #44715 [optimizer] introduce optimizer functional API, refactor Adagrad

Differential Revision: [D23935257](https://our.internmc.facebook.com/intern/diff/D23935257)

### CODE REVIEW

**Comment by vincentqb** [comment]
@egrefen @seba-1511  -- since you were part of the #39279 discussion, do you have any feedback on the current functional implementation?

File: torch/optim/adam.py [hunk 1]
```diff
@@ -47,6 +47,20 @@ def __init__(self, params, lr=1e-3, betas=(0.9, 0.999), eps=1e-8,
                         weight_decay=weight_decay, amsgrad=amsgrad)
         super(Adam, self).__init__(params, defaults)
 
+        # State initialization
+        for group in self.param_groups:
+            for p in group['params']:
+                state = self.state[p]
+                state['step'] = 0
+                # Exponential moving average of gradient values
+                state['exp_avg'] = torch.zeros_like(p, memory_format=torch.preserve_format)
```
Comment (albanD) [inline torch/optim/adam.py:56]: I don't think we want to move these to the init.
Some new features like [this](https://github.com/pytorch/pytorch/pull/44538#issuecomment-697170994) actually relies on the fact that our optimizers lazily initialize their buffers.

**Review by albanD — COMMENTED** [review]
Looks good.
Just the init that should be moved back to be lazy.

File: torch/optim/functional.py [hunk 2]
```diff
@@ -44,3 +46,51 @@ def make_sparse(values):
...
+        exp_avg_sq = exp_avg_sqs[i]
+        step = state_steps[i]
+        if amsgrad:
+            max_exp_avg_sq = max_exp_avg_sqs[i]
+
+        bias_correction1 = 1 - beta1 ** step
+        bias_correction2 = 1 - beta2 ** step
+
+        if weight_decay != 0:
+            grad = grad.add(p, alpha=weight_decay)
```
Comment (vincentqb) [inline torch/optim/functional.py:81]: `p` or `param`?

Comment (wanchaol) [reply torch/optim/functional.py]: seems like the `test_adam` in test_optim didn't cover the `weight_decay` case, added a test case to cover that.

**Review by vincentqb — APPROVED** [review]
LGTM!

**Pull request closed without merging**